README.md

```diff
@@ -4,105 +4,215 @@ Universum
+
+This tutorial has several parts:
+
+1. Philosophy and motivation.
+2. How to use it.
+3. What we've changed in `Prelude` (some gotchas).
+4. What already known things but that are not in `Prelude` we've brought into scope.
+5. What new thing we've added.
+
+This is not a tutorial on _Haskell_ or not even tutorial on each function. For detailed
+documentation of every function with examples and usages see
+[_Haddock documentation_](http://hackage.haskell.org/package/universum).
+
+Why another custom Prelude?
+---------------------------
+
+### Motivation
+
+In [Serokell](https://github.com/serokell/) we want to be as much productive as possible.
+That's why we are using [_Haskell_](https://haskell-lang.org/). This choice implies
+that we're restricted to use [`Prelude`](http://hackage.haskell.org/package/base-4.9.1.0/docs/Prelude.html):
+impicit import of basic functions, type classes and data type. But default `Prelude`
+[is considered to be not so good](https://news.ycombinator.com/item?id=8002749)
+due to some historical reasons.
+
+So we decided to use better tool. Hopefully _Haskell_ provides us with ability
+to replace default `Prelude` with some alternative. All we need to do is to implement
+new basic set of defaults. But we don't actually want to implement everything from scratch.
+There're plenty of [already created custom preludes](https://guide.aelve.com/haskell/alternative-preludes-zr69k1hc).
+After long and hot discussion our team decided to base our custom prelude on
+[`protolude`](https://github.com/sdiehl/protolude). If you're not familiar with it
+you can read [tutorial about `protolude`](http://www.stephendiehl.com/posts/protolude.html).
+
+Next section describes why we made this choice and what we're willing to do.
+This tutorial doesn't cover differences from `protolude`. Intstead it says what
+differs from custom `Prelude`.
+
+### Main goals motivation
+
+While creating and maintaining custom prelude we're following next goals:
+
+1. Avoid all [partial functions](https://www.reddit.com/r/haskell/comments/5n51u3/why_are_partial_functions_as_in_head_tail_bad/).
+   We like [total](http://mathworld.wolfram.com/TotalFunction.html) and exception-free functions.
+   You can still use some [_unsafe_](https://github.com/serokell/universum/blob/master/src/Unsafe.hs)
+   functions but they are not exported by default.
+2. Use more efficient [string representations](https://www.reddit.com/r/haskell/comments/29jw0s/whats_wrong_with_string/).
+   `String` type is crushingly inefficient. All our functions tries to be polymorphic over string
+   types or use [`Text`](http://hackage.haskell.org/package/text-1.2.2.1/docs/Data-Text.html)
+   as the default string type. Though `String` type alias is still reexported because community
+   is evolving slowly, some libraries still use `String` type. But we recommend to avoid `String`!
+3. Don't reinvent wheels. We're not trying to rebuild whole type hierarchy from scratch
```

Comment:
I mean, it's not reinventing wheels (strictly speaking), but it's similar to what `classy-prelude` does.